src/OrderedSet.mo

```diff
@@ -0,0 +1,1201 @@
+            #black(r2, x, r3)
+          )
+        };
+        case _ { Debug.trap "balRight" }
+      }
+    };
+
+    func append<T>(left : Tree<T>, right : Tree<T>) : Tree<T> {
+      switch (left, right) {
+        case (#leaf, _) { right };
+        case (_, #leaf) { left };
+        case (
+          #red(l1, x1, r1),
+          #red(l2, x2, r2)
+        ) {
+          switch (append(r1, l2)) {
+            case (#red(l3, x3, r3)) {
+              #red(
+                #red(l1, x1, l3),
+                x3,
+                #red(r3, x2, r2)
+              )
+            };
+            case r1l2 {
+              #red(l1, x1, #red(r1l2, x2, r2))
+            }
+          }
+        };
+        case (t1, #red(l2, x2, r2)) {
+          #red(append(t1, l2), x2, r2)
+        };
+        case (#red(l1, x1, r1), t2) {
+          #red(l1, x1, append(r1, t2))
+        };
+        case (#black(l1, x1, r1), #black(l2, x2, r2)) {
+          switch (append(r1, l2)) {
+            case (#red(l3, x3, r3)) {
+              #red(
+                #black(l1, x1, l3),
+                x3,
+                #black(r3, x2, r2)
+              )
+            };
+            case r1l2 {
+              balLeft(
+                l1,
+                x1,
+                #black(r1l2, x2, r2)
+              )
+            }
+          }
+        }
+      }
+    };
+
+    public func delete<T>(s : Set<T>, compare : (T, T) -> O.Order, x : T) : Set<T> {
+      var changed : Bool = false;
+      func delNode(left : Tree<T>, x1 : T, right : Tree<T>) : Tree<T> {
+        switch (compare(x, x1)) {
+          case (#less) {
+            let newLeft = del left;
+            switch left {
+              case (#black(_, _, _)) {
+                balLeft(newLeft, x1, right)
+              };
+              case _ {
+                #red(newLeft, x1, right)
+              }
+            }
+          };
+          case (#greater) {
+            let newRight = del right;
+            switch right {
+              case (#black(_, _, _)) {
+                balRight(left, x1, newRight)
+              };
+              case _ {
+                #red(left, x1, newRight)
+              }
+            }
+          };
+          case (#equal) {
+            changed := true;
+            append(left, right)
+          }
+        }
+      };
+      func del(tree : Tree<T>) : Tree<T> {
+        switch tree {
+          case (#black(left, x1, right)) {
+            delNode(left, x1, right)
+          };
+          case (#red(left, x1, right)) {
+            delNode(left, x1, right)
+          };
+          case (#leaf) {
+            tree
+          }
+        }
+      };
+      let newRoot = switch (del(s.root)) {
+        case (#red(left, x1, right)) {
+          #black(left, x1, right)
+        };
+        case other { other }
+      };
+      { root = newRoot;
+        size = if changed { s.size -1 } else { s.size } }
+    }
+  };
+
+  /// Test helpers
+  public module SetDebug {
```

Comment:
Should we move this module into the test dir, to avoid polluting the library?

Reply:
Ah, sorry, GH wasn't showing me your reply.... One could actually still hide but types are just structural and we have no type abstraction (sigh). You could literally redeclare tree or something equivalent and it would work.